### DISCUSSION TITLE
How to model my data suitable for GNNs?

### DISCUSSION THREAD

**Discussion by shahinghasemi** [pyg-team/pytorch_geometric#3589] [Q&A]
First of all Thank you for your great support and effort. I’m quite new to GNNs so my questions might seem basic. 
These are my data(all are binary matrices):
- `D-B` of shape `(d, b)`
- `B-B` of shape `(b, b) `
- `D-P` of shape `(d, p)`
- `D-S` of shape `(d, s)`

The final task is to predict the `D-B` matrix, here’s the considerations/questions for this task:

1. I don’t have a matrix as `D-D` of shape `(d, d)` to use it as adjacency matrix for GNN operators. Is there any solution to use GNN operators without adjacency matrix or any approach to simulate the adjacency matrix in this situation?
2. We can use `D-P` and `D-S` as feature matrices for nodes of type `D`. then How can I involve them as so? 
3. One solution I thought to approach the task, is to convert the problem from **link prediction** to **feature prediction**. This way we can use  `B-B`, and `(D-B).T` as adjacency and feature matrix respectively. However this way we’re ignoring `D-P` and `D-S` matrices which can provide further information to improve the accuracy(is there any way to involve them using this approach?).
4. Another approach I thought of was to consider the whole data as a hetergenous network with 4 types of nodes(`D`, `B`, `P`, `S`) and edges. However it’s not clear to me how should I handle the problem with this setup since I don’t have adjacency for each node type(except for `B`)?

Any suggestion is highly appreciated. Thanks in advance.

**Comment by rusty1s** [comment]
1. I'm not sure I understand. Is `D-B` an adjacency matrix connecting node types `D` and `B`, modeling a [heterogeneous graph](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html)? If you don't have an adjacency matrix, your best bet is to try to create them synthetically, e.g., via `torch_geometric.nn.radius_graph` or `torch_geometric.nn.knn_graph`.
2. You should be able to concatenate them together into `[D x P + S]`.
